### ISSUE TITLE
Overhaul JS Importer API

### ISSUE THREAD

**Issue by nex3** [sass/sass#2509]
* [x] [Proposal](https://github.com/sass/sass/blob/main/proposal/new-js-importer.d.ts)
* [x] [Tests](https://github.com/sass/sass-spec/pull/1727)
* [x] [Documentation](https://github.com/sass/sass/pull/3183)
* [x] [Dart Sass](https://github.com/sass/dart-sass/pull/1522)
* [x] Node Host

---

The JavaScript importer API shared by Dart Sass and Node Sass needs an overhaul. The current API has a number of issues:

* It doesn't respect URL semantics. In particular, it doesn't provide any guarantee that `@import "./bar"` within a stylesheet with URL `foo` means the same thing as `@import "foo/bar"`, even when resolved by the same importer.

* It doesn't have a notion of resolving an imported URL to an absolute URL—for example, an npm-aware importer might resolve `~foo` to `file:///home/nex3/app/node_modules/foo/_index.scss`. This makes canonicalization of imported stylesheets impossible, and means that importers are fundamentally less expressive than built-in filesystem imports. It also means that the URLs reported in source maps and other APIs are unlikely to be resolvable.

* It doesn't support the indented syntax.

* Its notion of "redirecting" to a file import has some strange corner cases—for example, a relative path redirect can be resolved relative to the file that contains the import, relative to the working directory, *or* relative to an import path.

* It doesn't seem to do much error checking for invalid responses.

* It uses the string `"stdin"` to refer to files passed by data, even when they don't come from standard input. This could potentially conflict with a real file named `"stdin"`.

* The `this` context includes a bunch of undocumented information without a clear use-case.

We should design a new API that addresses these issues and makes it easy to write importers that work similarly to native filesystem imports. I strongly favor writing the API in such a way that filesystem imports themselves can be modeled as an importer.

**Comment by ArmorDarks** [comment]
I hope that `~foo` is just an example, since `~` isn't part of the NPM loader, but a something that people came up while doing custom Webpack loaders.

**Comment by nex3** [comment]
Yes, it's just an example.

**Comment by nex3** [comment]
This proposal is now officially out for review. It will remain that way for at least a month (potentially longer depending on discussion), at which point we'll move on to implementation.

**Comment by nex3** [comment]
It's been a month and there have been no objections, so I'm going to mark the proposal as accepted and start merging it into the main spec directory.

**Comment by nex3** [comment]
This is now usable in Dart Sass 1.45.0-rc.1! We'd love people's feedback on it before we launch a full stable release some time next week.

**Comment by clshortfuse** [comment]
@nex3 Hi!

I'm probably one of your projects, uh, more advanced users, and been using a personal fork for the past year because of a need to support code-splitting with Webpack chunks within a monorepo handled by a yarn workspace (mouthful). I'm trying a bit to understand the new JS importer in hopes I can move back to to upstream. I'm not sure how we can canonicalize URLs without knowing what file is requesting it.

For example, given 

````js
sass.compileString('@use material/card;', {
  url: new URL(path.resolve('./index.scss'), 'file://');
  importer: {
    canonicalize(url) { /* ... */ }
    load(canonicalURL) { /* ... */ }
   }
}
````

The URLs that need to be canonlicalized in order are:

* `@material/card/_index.scss`
  * Found `file:///home/carlos/projects/sass-importer-test/node_modules/@material/card/_index.scss`
* `file:///home/carlos/projects/sass-importer-test/node_modules/@material/card/variables`
  *  Found `file:///home/carlos/projects/sass-importer-test/node_modules/@material/card/_variables.scss`
* `file:///home/carlos/projects/sass-importer-test/node_modules/@material/card/@material/theme/theme`
  * Cannot be found

For the third one, if I know what file is requesting the URL over the `@use` command, then I can resolve it. But without it, it's impossible. To be honest, I tried hacking it, forcing it to be synchronous and trying to trace client-side what file `sass` is working on by looking at the last request:

* Canonicalize: `file:///home/carlos/projects/sass-importer-test/@material/card/_index.scss`
  * Found `/home/carlos/projects/sass-importer-test/node_modules/@material/card/_index.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/card/variables`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/card/_variables.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/card/@material/theme/theme`
  * Found `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_theme.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/@material/feature-targeting/feature-targeting`
  * Found `/home/carlos/projects/sass-importer-test/node_modules/@material/feature-targeting/_feature-targeting.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/css`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_css.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/gss`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_gss.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/custom-properties`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_custom-properties.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/css`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_css.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/gss`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_gss.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/gss`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_gss.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/keys`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_keys.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/custom-properties`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_custom-properties.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/replace`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_replace.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/theme-color`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_theme-color.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/custom-properties`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_custom-properties.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/theme/keys`
  * Resolved `/home/carlos/projects/sass-importer-test/node_modules/@material/theme/_keys.scss`
* Canonicalize: `file:///home/carlos/projects/sass-importer-test/node_modules/@material/card/@material/theme/theme-color`
  * Cannot be found 

It doesn't really work because I can't capture that SASS popped back up to `@material/card/_index.scss`. It seems the importer is not low level enough to let me handle how files are imported, resolved, and cached.

I guess this has to 100% work with absolute paths, but that gets pretty difficult to manage with workspaces. I'm going to ping https://github.com/sass/sass/issues/2739 since it seems this would be related.

Note: This isn't my actual workflow, but just an example of the limitations of the new importer.

**Comment by nex3** [comment]
**tl;dr:** You need to pass your importer to the [`importers`](https://sass-lang.com/documentation/js-api/interfaces/Options#importers) option as well if you want to use it for non-relative imports. If you're always importing things from the filesystem (which it looks like you are), it's also a good idea to use a [`FileImporter`](https://sass-lang.com/documentation/js-api/interfaces/FileImporter).

---

To give you more insight into what's happening under the covers here: the new API expects all importers' URL spaces to be basically path-like. This means that a relative load is always resolved relative to the original file's canonical URL, and a file that's loaded from a relative URL can always also be loaded from an absolute URL as well. To ensure this, the API handes relative loads as follows:

* First, it resolves the imported URL relative to the current file's canonical URL. In your example, this produces `"file:///home/carlos/projects/sass-importer-test/node_modules/@material/card/@material/theme/theme"`.

* Then, it passes the result to the current file's importer. This ensures that the importer won't ever return a different result for relative loads than it would for absolute loads.

* An importer's `canonicalize()` and `load()` functions always have the option to return `null` to say "I don't recognize this" or "this file doesn't actually exist". If the current file's importer returns `null`, the API continues:

* For each importer in `importers`, the API passes the *original* URL (`"@material/theme/theme"`) to that importer's `canonicalize()` method. (This is why `canonicalize()` takes a string rather than a `URL`: so it can receive these relative URLs.) The first one to return a non-`null` value loads the file.

So you see, the importer you pass in to [`StringOptions.importer`](https://sass-lang.com/documentation/js-api/interfaces/StringOptionsWithImporter#importer) is *only* used for relative loads. As soon as you want to load something that's not relative to the current stylesheet, you need to use the `importers` option instead.
